File: js/cracking_coding_interview/trees_graphs/successor.js
function successor(node){
  if (!node) {
    return null;
  }
  if (node.right){
    return leftMost(node.right);
  } else {
    while (node.parent && node === node.parent.right) {
      node = node.parent;
    }
    return node.parent;
  }
}

function leftMost(node){
  if (!node) {
    return null;
  }
  while (node.left !== null) {
    node = node.left;
  }
  return node;
}

File: ruby/cracking_coding_interview/trees_graphs/successor.rb
# Write an algorithm to find the "next" node (i.e., in-order successor) of a given node in an binary search tree. You may assume that each node has a link to its parent.

def successor(node)
  return nil if node.nil?

  if node.right.nil?
    succ = node.parent
    while succ
      if succ.val > node.val
        return succ
      end
      succ = succ.parent
    end
    return nil
  else
    succ = node.right
    while succ.left
      succ = succ.left
    end
    return succ
  end
end

# time: logn
# space: 1
def successor2(node)
  return nil if node.nil?

  if node.right
    succ = node.right
    while succ.left
      succ = succ.left
    end
    return succ
  else
    succ = node.parent
    while succ && succ.left != node
      succ = succ.parent
      node = succ
    end
    return succ
  end
end
